### DISCUSSION TITLE
What's in a Good Error Message? - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#291] [Announcements]
# What's in a Good Error Message? - Gunnar Morling

As software developers, we’ve all come across those annoying, not-so-useful error messages when using some library or framework: "Couldn’t parse config file", "Lacking permission for this operation", etc.
Ok, ok, so something went wrong apparently; but what exactly? What config file? Which permissions? And what should you do about it?
Error messages lacking this kind of information quickly create a feeling of frustration and helplessness.

So what makes a good error message then?
To me, it boils down to three pieces of information which should be conveyed by an error message:

Context: What led to the error? What was the code trying to do when it failed?

The error itself: What exactly failed?

Mitigation: What needs to be done in order to overcome the error?

https://www.morling.dev/blog/whats-in-a-good-error-message/

**Comment by RossVertizan** [comment]
Excellent article which reflects my personal approach to meaningful error messages.

One thing that I would add is an indication of severity.  Personally, I like to have every message preceded with a word which indicates the severity of the message, for example:

INFO - Indicates that the message is for information only, everything is working as expected and this is for politeness and usability only. e.g INFO: Config file <xxx> parsed successfully.

NOTICE - Indicates that the message is alerting the user to something that is not going to cause a problem and the code can continue to run as normal, but it might not be what would be normally expected e.g. NOTICE: Config file <xxx> contains no settings

WARNING - Indicates that the message is alerting the user to something that is wrong, the code will continue to run but the results might not be what you would expect e.g. WARNING:  Config file <xxx> contains repeated settings [paramA, paramD], using values from last entries in file.

ERROR - Indicates that something has gone wrong and anything after this point should be ignored e.g. ERROR: Couldn’t parse config file: /etc/sample-config.properties; given snapshot mode 'nevr' isn’t valid (must be one of 'initial', 'always', 'never')

and then I always have 

DEBUG - Indicates that the message is for debug purposes only.

I would always implement a verbosity switch to allow different levels of error message.  You don't always need DEBUG and INFO messages for example.

Having these clearly defined severity levels allows you to quickly home in on the cause of issues.

Of course this is all outside the scope of your article because you specifically, identify it is about error messages, so please forgive me for this diversion.

**Reply by vphilipnyc** [reply]
Yes!  During development, I love seeing things like “must be one of 'initial', 'always', 'never'.”  Gives me confidence in the codebase.

**Reply by gunnarmorling** [reply]
Thanks for the comment, @RossVertizan! Totally agreed that adding a severity makes sense, luckily logging frameworks take care of that. Very happy to hear you found this post useful.

**Comment by a-karro** [comment]
OTOH let's not forget who's going to see the error message: https://owasp.org/www-community/Improper_Error_Handling

**Reply by gunnarmorling** [reply]
Totally agreed on the importance of this; this article is not about user-facing error messages, though.

**Comment by toledoroy** [comment]
Can't love this enought
I'm so glad to have finally found another developer that appreciates the value of a good error message.

**Reply by gunnarmorling** [reply]
Ha, thanks! You're not alone :)

**Reply by toledoroy** [reply]
I also like to dump relevant state variables and function input in the error message. Makes debugging a breeze.
What do you think, any reason not to?

**Comment by PeterAdam** [comment]
And as the recent log4j fiasco told us, there are very good use cases for including too much in the logging system!

**Reply by zmauzmau** [reply]
Great article
Will forward.
The real question is : will it help :-)

**Comment by bjthomson** [comment]
And never forget that all user input is assumed to be evil. When logging user input it must be sanitised / escaped, since the developer viewing the log is probably running at elevated privilege. That was a flaw in a logging framework popular in the late nineties / early noughties called "Analog" if memory serves.
